README.md
# Game of Life parsers

[![Build Status](https://travis-ci.org/Leopard2A5/game-of-life-parsers.svg?branch=master)](https://travis-ci.org/Leopard2A5/game-of-life-parsers)
[Documentation](https://docs.rs/game-of-life-parsers/)

Collection of parsers for Conway's game of life.

Currently supported file formats:

* Life 1.05
* Life 1.06

## Usage
```rust
extern crate game_of_life_parsers;
// use std::fs::File;
use game_of_life_parsers::{Parser, Life105Parser};

fn main() {
    // let file = File::open("file.life").unwrap();
    let file = "#N\n#P 0 0\n..*".as_bytes();
    let mut parser = Life105Parser::new();
    let game_descriptor = parser.parse(Box::new(file)).unwrap();
    for live_cell in game_descriptor.live_cells() {
        // iterate overe live cells
    }
}
```


CHANGELOG.md
# Changelog

## 2.0.1
* Now raising `errors::ErrorKind::MalformedLine` on all unexpected chars
* Ignoring empty and commented out lines

## 2.0.0
* Improved handling of IO errors
* Parsers now raise error when encountering format tag of different format
* Added this changelog
* Fixed error with malformed block header lines in life 1.05 files
* Added line number to `errors::ErrorKind::MalformedLine`
* Removed `errors::ErrorKind::InvalidRulesLine` in favor of `errors::ErrorKind::MalformedLine`
* Now correctly handling block offsets in 1.05 parser and cell coordinates in 1.06 parser that don't fit into an `i16`. Both will raise a `errors::ErrorKind::CoordinateOutOfRange`
* Life 1.05 parser now raises `MalformedLine` on unexpected characters inside blocks
* Life 1.05 parser now correctly handles coordinates that don't fit into an `i16`

## 1.0.0
* Input to parsers is now boxed, to make `Parser` a proper trait object

## 0.2.0
* Add parser for life 1.06 files


tests/test_life_1_05.rs
extern crate game_of_life_parsers;

use std::fs::File;
use game_of_life_parsers::GameDescriptor;
use game_of_life_parsers::Coord;
use game_of_life_parsers::Parser;
use game_of_life_parsers::Life105Parser;

#[test]
fn parse_file() {
	let file = File::open("tests/life_1_05/glider.life").unwrap();
	let mut parser = Life105Parser::new();

	let gd: Box<GameDescriptor> = parser.parse(Box::new(file)).unwrap();
	assert_eq!(&[2, 3], gd.survival());
	assert_eq!(&[1], gd.birth());
	assert_eq!(&[
		// block 1
		Coord { x:  0, y: -1 },
		Coord { x:  1, y: 0 },
		Coord { x: -1, y: 1 }, Coord { x: 0, y: 1 }, Coord { x: 1, y: 1 },
		// block 2
		Coord { x: 3, y: 2 }, Coord { x: 4, y: 3 }, Coord { x: 5, y: 4 }
	], gd.live_cells());
}


tests/test_life_1_06.rs
extern crate game_of_life_parsers;

use std::fs::File;
use game_of_life_parsers::Coord;
use game_of_life_parsers::Parser;
use game_of_life_parsers::Life106Parser;

#[test]
fn parse_file() {
	let file = File::open("tests/life_1_06/test_1_06.life").unwrap();
	let mut parser = Life106Parser::new();

	let gd = parser.parse(Box::new(file)).unwrap();
	assert_eq!(&[
		Coord { x: 0, y: 0 },
		Coord { x: 1, y: 0 },
		Coord { x: 3, y: 5 }
	], gd.live_cells())
}


src/lib.rs
#![recursion_limit = "1024"]

//! Collection of parsers for Conway's game of life.
//!
//! Currently supported file formats:
//!
//! * Life 1.05
//!
//! ## Usage
//!
//! ```
//! extern crate game_of_life_parsers;
//! // use std::fs::File;
//! use game_of_life_parsers::{Parser, Life105Parser};
//!
//! fn main() {
//! 	// let file = File::open("file.life").unwrap();
//! 	let file = "#N\n#P 0 0\n..*".as_bytes();
//! 	let mut parser = Life105Parser::new();
//! 	let game_descriptor =  parser.parse(Box::new(file)).unwrap();
//! 	for live_cell in game_descriptor.live_cells() {
//! 		// iterate overe live cells
//!		}
//! }
//! ```

#[macro_use]
extern crate error_chain;
extern crate regex;

/// Error types generated by `error_chain`.
pub mod errors {
	use std::io;

	error_chain! {
		errors {
			IOError(error: io::ErrorKind)
			InvalidFileFormat
			MalformedLine(line: usize)
			CoordinateOutOfRange(line: usize)
		}
	}
}

mod game_descriptor;
mod default_game_descriptor;
mod parsers;

pub use game_descriptor::GameDescriptor;
pub use game_descriptor::Coord;
pub use parsers::Parser;
pub use parsers::Life105Parser;
pub use parsers::Life106Parser;


src/stringreader.rs
use std::io::Read;
use std::io::Result;
use std::slice::Iter;

pub struct StringReader<'a> {
	iter: Iter<'a, u8>,
}

impl<'a> StringReader<'a> {
	pub fn new(data: &'a str) -> Self {
		Self {
			iter: data.as_bytes().iter(),
		}
	}
}

impl<'a> Read for StringReader<'a> {
	fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
		for i in 0..buf.len() {
			if let Some(x) = self.iter.next() {
				buf[i] = *x;
			} else {
				return Ok(i);
			}
		}
		Ok(buf.len())
	}
}

#[cfg(test)]
mod test {
	use super::*;
	use std::io::{BufRead, BufReader};

	#[test]
	fn test() {
		let data = "abc\ndef";
		let mut reader = BufReader::new(StringReader::new(data));
		let mut buffer = String::new();

		buffer.clear();
		reader.read_line(&mut buffer).unwrap();
		assert_eq!("abc\n", buffer);

		buffer.clear();
		reader.read_line(&mut buffer).unwrap();
		assert_eq!("def", buffer);
	}
}


src/default_game_descriptor.rs
use ::GameDescriptor;
use ::Coord;

pub struct DefaultGameDescriptor {
	survival: Vec<u8>,
	birth: Vec<u8>,
	live_cells: Vec<Coord>,
}

impl DefaultGameDescriptor {
	pub fn new() -> Self {
		DefaultGameDescriptor {
			survival: vec![],
			birth: vec![],
			live_cells: vec![],
		}
	}

	pub fn add_survival(&mut self, value: u8) {
		self.survival.push(value);
	}

	pub fn add_birth(&mut self, value: u8) {
		self.birth.push(value);
	}

	pub fn clear_rules(&mut self) {
		self.survival.clear();
		self.birth.clear();
	}

	pub fn add_live_cell(&mut self, x: i16, y: i16) {
		self.live_cells.push(Coord{ x: x, y: y });
	}
}

impl GameDescriptor for DefaultGameDescriptor {
	fn survival(&self) -> &[u8] {
		&self.survival
	}

	fn birth(&self) -> &[u8] {
		&self.birth
	}

	fn live_cells(&self) -> &[Coord] {
		&self.live_cells
	}

	fn no_negative_coords(&self) -> Box<GameDescriptor> {
		use std::cmp;

		let min_x = self.live_cells.iter()
			.map(|item| item.x)
			.min()
			.unwrap_or(0);
		let min_x = cmp::min(0, min_x);

		let min_y = self.live_cells.iter()
			.map(|item| item.y)
			.min()
			.unwrap_or(0);
		let min_y = cmp::min(0, min_y);

		let mut ret = DefaultGameDescriptor::new();
		for survival in &self.survival {
			ret.add_survival(*survival);
		}
		for birth in &self.birth {
			ret.add_birth(*birth);
		}

		for coord in &self.live_cells {
			ret.add_live_cell(
				coord.x - min_x,
				coord.y - min_y
			)
		}

		Box::new(ret)
	}
}

#[cfg(test)]
mod test {
	use ::Life105Parser;
	use ::Parser;
	use ::Coord;

	#[test]
	fn no_negative_coords_should_work() {
		let input = r#"
#N
#P 0 -2
.*.
#P -1 0
*..
"#;
		let mut parser = Life105Parser::new();
		let gd = parser.parse(Box::new(input.as_bytes()))
			.unwrap();
		assert_eq!(&[ Coord {x: 1, y: -2}, Coord {x: -1, y: 0} ], gd.live_cells());

		let gd = gd.no_negative_coords();
		assert_eq!(&[ Coord {x: 2, y: 0}, Coord {x: 0, y: 2} ], gd.live_cells());
	}
}


src/parsers/life_1_06_parser.rs
use super::Parser;
use ::errors::{self, Error, ErrorKind};
use ::GameDescriptor;
use ::default_game_descriptor::DefaultGameDescriptor;
use std::io::{Read, BufRead, BufReader};
use regex::Regex;

/// Parser for files in the life 1.05 format.
pub struct Life106Parser {}

impl Life106Parser {
	/// Construct a new instance.
	pub fn new() -> Self {
		Life106Parser{}
	}
}

impl Parser for Life106Parser {
	fn parse<'a>(&mut self, input: Box<Read + 'a>) -> errors::Result<Box<GameDescriptor>> {
		let reader = BufReader::new(input);
		let mut ret = DefaultGameDescriptor::new();
		let regex = Regex::new("(\\d+)\\s+(\\d+)")
			.expect("invalid regex!");

		for (line_num, line) in reader.lines().enumerate() {
			let line_num = line_num + 1;
			let line = line.
				map_err(|err| errors::ErrorKind::IOError(err.kind()))?;
			let line = line.trim();

			if line.len() == 0 {
				// ignore
			} else if line.starts_with("#Life") {
				check_file_format(line)?;
			} else if regex.is_match(line) {
				let coords: Vec<&str> = line.split_whitespace().collect();
				let x = coords[0].parse::<i16>()
					.map_err(|_| Error::from(ErrorKind::CoordinateOutOfRange(line_num)))?;
				let y = coords[1].parse::<i16>()
					.map_err(|_| Error::from(ErrorKind::CoordinateOutOfRange(line_num)))?;

				ret.add_live_cell(x, y);
			} else if line.starts_with("#") {
				// ignore
			} else {
				bail!(ErrorKind::MalformedLine(line_num))
			}
		}

		Ok(Box::new(ret))
	}
}

fn check_file_format(line: &str) -> errors::Result<()> {
	use regex::Regex;

	let regex = Regex::new("#Life\\s+1.06").expect("Invalid regex");
	if regex.is_match(line) {
		Ok(())
	} else {
		bail!(ErrorKind::InvalidFileFormat)
	}
}

#[cfg(test)]
mod test {
	extern crate io_test_util;

	use super::*;
	use std::io;
	use errors::Error;
	use errors::ErrorKind::*;

	#[test]
	fn should_correctly_handle_io_errors() {
		use self::io_test_util;

		let mut parser = Life106Parser::new();
		let reader = io_test_util::ErrReader::new(io::ErrorKind::NotFound);

		match parser.parse(Box::new(reader)) {
			Err(errors::Error(errors::ErrorKind::IOError(io::ErrorKind::NotFound), _)) => {},
			Err(_) => panic!("Wrong error returned!"),
			Ok(_) => panic!("No error returned!")
		}
	}

	#[test]
	fn should_throw_error_on_wrong_format_annotation() {
		let mut parser = Life106Parser::new();
		let input = Box::new("#Life 1.05\n#P -1 -1\n.*.".as_bytes());
		let res = parser.parse(input);
		match res {
			Err(errors::Error(errors::ErrorKind::InvalidFileFormat, _)) => {},
			Err(errors::Error(x, _)) => panic!("Unexpected error {}", x),
			_ => panic!("No error thrown!"),

		}
	}

	#[test]
	fn should_handle_too_big_coordinates() {
		let mut parser = Life106Parser::new();
		let input = Box::new("32768 0".as_bytes());
		match parser.parse(input) {
			Err(Error(CoordinateOutOfRange(1), _)) => {},
			Err(_) => panic!("Wrong error thrown!"),
			_ => panic!("No error thrown!")
		}
	}

	#[test]
	fn should_fail_on_unexpected_chars() {
		let input = Box::new("0 0\n1 1 \nYo Yo!".as_bytes());
		let mut parser = Life106Parser::new();
		match parser.parse(input) {
			Err(Error(MalformedLine(3), _)) => {},
			_ => panic!("Expected MalformedLine!")
		}
	}

	#[test]
	fn should_ignore_empty_lines() {
		let input = Box::new("\n\n0 0\n1 1".as_bytes());
		let mut parser = Life106Parser::new();
		match parser.parse(input) {
			Ok(_) => {},
			_ => panic!("Expected Ok!")
		}
	}

	#[test]
	fn should_ignore_commented_out_lines() {
		let input = Box::new("0 0\n1 1 \n# Yo Yo!".as_bytes());
		let mut parser = Life106Parser::new();
		match parser.parse(input) {
			Ok(_) => {},
			_ => panic!("Expected Ok!")
		}
	}
}


src/parsers/life_1_05_parser.rs
use std::io::{BufRead, BufReader, Read};
use super::Parser;
use ::errors::{self, Error, ErrorKind};
use ::GameDescriptor;
use ::default_game_descriptor::DefaultGameDescriptor;
use std::i16;

/// Parser for files in the life 1.05 format.
pub struct Life105Parser {}

impl Life105Parser {
	/// Construct a new instance.
	pub fn new() -> Self {
		Life105Parser{}
	}
}

impl Parser for Life105Parser {
	fn parse<'a>(&mut self, input: Box<Read + 'a>) -> errors::Result<Box<GameDescriptor>> {
		let mut ret = DefaultGameDescriptor::new();

		let mut offset = None;
		let mut line_in_block: i16 = 0;
		for (line_num, line) in BufReader::new(input).lines().enumerate() {
			let line_num = line_num + 1; // line numbers don't start at 0
			let line = line
				.map_err(|err| errors::ErrorKind::IOError(err.kind()))?;
			let line = line.trim();

			if line.len() == 0 {
				// ignore
			} else if line.starts_with("#R") {
				ret.clear_rules();
				parse_rules(line_num, &line, &mut ret)?;
			} else if line.starts_with("#Life") {
				check_file_format(line)?;
			} else if line.starts_with("#N") {
				ret.clear_rules();
				ret.add_survival(2);
				ret.add_survival(3);
				ret.add_birth(3);
			} else if line.starts_with("#P") {
				offset = Some(parse_offset(line_num, &line)?);
				line_in_block = 0;
			} else if let Some((ox, oy)) = offset {
				for (index, char) in line.chars().enumerate() {
					let index_with_offset = index as i32 + ox as i32;
					if index_with_offset > i16::MAX as i32 {
						bail!(ErrorKind::CoordinateOutOfRange(line_num));
					}
					let line_in_block_with_offset = line_in_block as i32 + oy as i32;
					if line_in_block_with_offset > i16::MAX as i32 {
						bail!(ErrorKind::CoordinateOutOfRange(line_num));
					}
					match char {
						'*' => ret.add_live_cell(index as i16 + ox, line_in_block + oy),
						'.' => {},
						_ => bail!(ErrorKind::MalformedLine(line_num))
					}
				}
				line_in_block += 1;
			} else if line.starts_with("#") {
				// ignore
			} else {
				bail!(ErrorKind::MalformedLine(line_num))
			}
		}

		Ok(Box::new(ret))
	}
}

fn parse_rules(
	line_num: usize,
	line: &str,
	gd: &mut DefaultGameDescriptor
) -> errors::Result<()> {
	use regex::Regex;

	let regex = Regex::new("#R\\s*(\\d+)\\s*/\\s*(\\d+)\\s*")
		.expect("invalid regex!");

	if let Some(captures) = regex.captures(line) {
		let survival = captures.get(1)
			.expect("Failed to get regex group")
			.as_str();
		let birth = captures.get(2)
			.expect("Failed to get regex group")
			.as_str();

		for char in survival.chars() {
			let digit = char.to_digit(10)
				.expect("Failed to parse u8") as u8;
			gd.add_survival(digit);
		}

		for char in birth.chars() {
			let digit = char.to_digit(10)
				.expect("Failed to parse u8") as u8;
			gd.add_birth(digit);
		}

		Ok(())
	} else {
		bail!(ErrorKind::MalformedLine(line_num))
	}
}

fn parse_offset(
	line_num: usize,
	line: &str,
) -> errors::Result<(i16, i16)> {
	use regex::Regex;

	let regex = Regex::new("#P\\s*([+-]?\\d+)\\s*([+-]?\\d+)\\s*").expect("Invalid regex");
	if let Some(captures) = regex.captures(line) {
		let x = captures.get(1).unwrap().as_str();
		let y = captures.get(2).unwrap().as_str();

		let x = x.parse::<i16>()
			.map_err(|_| Error::from(ErrorKind::CoordinateOutOfRange(line_num)))?;
		let y = y.parse::<i16>()
			.map_err(|_| Error::from(ErrorKind::CoordinateOutOfRange(line_num)))?;

		Ok((x, y))
	} else {
		bail!(errors::ErrorKind::MalformedLine(line_num))
	}
}

fn check_file_format(line: &str) -> errors::Result<()> {
	use regex::Regex;

	let regex = Regex::new("#Life\\s+1.05").expect("Invalid regex");
	if regex.is_match(line) {
		Ok(())
	} else {
		bail!(ErrorKind::InvalidFileFormat)
	}
}

#[cfg(test)]
mod test {
	extern crate io_test_util;

	use super::*;
	use errors::Error;
	use errors::ErrorKind::*;
	use std::io;

	#[test]
	fn parse_rules_should_err() {
		let mut gd = DefaultGameDescriptor::new();
		match parse_rules(5, "#R 23", &mut gd) {
			Err(Error(MalformedLine(5), _)) => {},
			Err(_) => panic!("Wrong error thrown!"),
			_ => panic!("No error thrown!")
		}
	}

	#[test]
	fn parse_rules_should_parse_custom_rules() {
		let mut gd = DefaultGameDescriptor::new();
		parse_rules(1, "#R24 / 1", &mut gd).unwrap();
		assert_eq!(&[2, 4], gd.survival());
		assert_eq!(&[1], gd.birth());
	}

	#[test]
	fn parser_should_understand_default_rules() {
		let mut parser = Life105Parser::new();
		let gd = parser.parse(Box::new("#N".as_bytes())).unwrap();
		assert_eq!(&[2, 3], gd.survival());
		assert_eq!(&[3], gd.birth());
	}

	#[test]
	fn should_correctly_handle_io_errors() {
		use self::io_test_util;

		let mut parser = Life105Parser::new();
		let reader = io_test_util::ErrReader::new(io::ErrorKind::BrokenPipe);

		match parser.parse(Box::new(reader)) {
			Err(Error(IOError(io::ErrorKind::BrokenPipe), _)) => {},
			Err(_) => panic!("Wrong error returned!"),
			_ => panic!("No error returned!")
		}
	}

	#[test]
	fn should_throw_error_on_wrong_format_annotation() {
		let mut parser = Life105Parser::new();
		let input = Box::new("#Life 1.06\n0 0\n1 2".as_bytes());
		let res = parser.parse(input);
		match res {
			Err(Error(InvalidFileFormat, _)) => {},
			Err(Error(x, _)) => panic!("Unexpected error {}", x),
			_ => panic!("No error thrown!"),

		}
	}

	#[test]
	fn should_raise_error_on_invalid_block_header() {
		let mut parser = Life105Parser::new();
		let input = "#P 0 a".as_bytes();
		match parser.parse(Box::new(input)) {
			Err(Error(MalformedLine(1), _)) => {},
			Err(_) => panic!("Wrong error thrown!"),
			_ => panic!("No error thrown!")
		}
	}

	#[test]
	fn parse_offset_should_handle_too_big_coords() {
		match parse_offset(1, "#P 0 32768") {
			Err(Error(CoordinateOutOfRange(1), _)) => {},
			Err(_) => panic!("Wrong error thrown!"),
			Ok(_) => panic!("No error thrown!")
		}
	}

	#[test]
	fn should_raise_error_on_unexpected_character_in_block() {
		let input = Box::new("#P -1 -1\n..*#".as_bytes());
		let mut parser = Life105Parser::new();
		match parser.parse(input) {
			Err(Error(MalformedLine(2), _)) => {},
			_ => panic!("Expected MalformedLine")
		}
	}

	#[test]
	fn should_handle_too_big_horizontal_coords() {
		let input = Box::new("#P 32767 0\n**".as_bytes());

		let mut parser = Life105Parser::new();
		let result = parser.parse(input);
		match result {
			Err(Error(CoordinateOutOfRange(2), _)) => {},
			_ => panic!("Expected CoordinateOutOfRange")
		}
	}

	#[test]
	fn should_handle_too_big_vertical_coords() {
		let input = Box::new("#P 0 32767\n.\n*".as_bytes());

		let mut parser = Life105Parser::new();
		let result = parser.parse(input);
		match result {
			Err(Error(CoordinateOutOfRange(2), _)) => {},
			Err(x) => println!("XXX {:?}", x),
			_ => panic!("Expected CoordinateOutOfRange")
		}
	}

	#[test]
	fn should_fail_on_unexpected_chars_outside_blocks() {
		let input = Box::new("this is not a life file".as_bytes());
		let mut parser = Life105Parser::new();
		match parser.parse(input) {
			Err(Error(MalformedLine(1), _)) => {},
			_ => panic!("Expected MalformedLine!")
		}
	}

	#[test]
	fn empty_lines_should_be_ignored() {
		let text = r#"
#N

#P -1 -1
.*.
"#;
		let input = Box::new(text.as_bytes());
		let mut parser = Life105Parser::new();
		match parser.parse(input) {
			Ok(_) => {},
			_ => panic!("Expected Ok!")
		}
	}

	#[test]
	fn commented_out_lines_should_be_ignored() {
		let text = "# I am not a malformed line";
		let input = Box::new(text.as_bytes());
		let mut parser = Life105Parser::new();
		match parser.parse(input) {
			Ok(_) => {},
			_ => panic!("Expected Ok!")
		}
	}
}
